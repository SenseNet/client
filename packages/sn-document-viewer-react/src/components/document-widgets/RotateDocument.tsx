import IconButton from '@material-ui/core/IconButton'
import RotateLeft from '@material-ui/icons/RotateLeft'
import RotateRight from '@material-ui/icons/RotateRight'
import React, { useCallback } from 'react'
import { applyShapeRotations, ImageUtil } from '../..'
import { useDocumentData, useLocalization, usePreviewImages, useViewerState } from '../../hooks'
import { ROTATION_AMOUNT, ROTATION_MODE } from '../../models/RotationModel'
/**
 * Component that allows document rotation
 */
export const RotateDocumentWidget: React.FC<{ mode?: ROTATION_MODE }> = (props) => {
  const localization = useLocalization()
  const viewerState = useViewerState()
  const previewImages = usePreviewImages()
  const { documentData, updateDocumentData } = useDocumentData()

  const rotateDocument = useCallback(
    (direction: string) => {
      const newRotation = viewerState.rotation ?? []
      previewImages.imageData.forEach((page) => {
        const existingObj = newRotation.find((rotation) => rotation.pageNum === page.Index)
        if (newRotation.length > 0 && existingObj) {
          const prevValue = existingObj.degree || 0
          existingObj.degree = ImageUtil.normalizeDegrees(
            (prevValue + ((direction === 'left' ? -ROTATION_AMOUNT : ROTATION_AMOUNT) % 360)) % 360,
          )
        } else {
          newRotation.push({
            pageNum: page.Index,
            degree: ImageUtil.normalizeDegrees(
              (0 + ((direction === 'left' ? -ROTATION_AMOUNT : ROTATION_AMOUNT) % 360)) % 360,
            ),
          })
        }
      })

      viewerState.updateState({
        rotation: newRotation,
      })

      //update shapes as well
      const newImages = previewImages.imageData.map((img) => {
        updateDocumentData({
          shapes: {
            annotations: applyShapeRotations(
              documentData.shapes.annotations,
              direction === 'left' ? -ROTATION_AMOUNT : ROTATION_AMOUNT,
              img,
            ),
            highlights: applyShapeRotations(
              documentData.shapes.highlights,
              direction === 'left' ? -ROTATION_AMOUNT : ROTATION_AMOUNT,
              img,
            ),
            redactions: applyShapeRotations(
              documentData.shapes.redactions,
              direction === 'left' ? -ROTATION_AMOUNT : ROTATION_AMOUNT,
              img,
            ),
          },
        })

        return img
      })
      previewImages.setImageData(newImages)
    },
    [
      documentData.shapes.annotations,
      documentData.shapes.highlights,
      documentData.shapes.redactions,
      previewImages,
      updateDocumentData,
      viewerState,
    ],
  )

  const button = (direction: ROTATION_MODE) => {
    const isLeft = direction === ROTATION_MODE.anticlockwise
    return (
      <IconButton
        title={isLeft ? localization.rotateDocumentLeft : localization.rotateDocumentRight}
        onClick={isLeft ? () => rotateDocument('left') : () => rotateDocument('right')}
        id={isLeft ? 'RotateLeft' : 'RotateRight'}>
        {isLeft ? (
          <RotateLeft style={{ border: '2px solid', borderRadius: '5px' }} />
        ) : (
          <RotateRight style={{ border: '2px solid', borderRadius: '5px' }} />
        )}
      </IconButton>
    )
  }

  switch (props.mode) {
    case ROTATION_MODE.anticlockwise:
    case ROTATION_MODE.clockwise:
      return button(props.mode)
    default:
      return (
        <div style={{ display: 'inline-block' }}>
          {button(ROTATION_MODE.anticlockwise)}
          {button(ROTATION_MODE.clockwise)}
        </div>
      )
  }
}
