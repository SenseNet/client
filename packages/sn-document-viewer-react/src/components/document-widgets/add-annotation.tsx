import { createStyles, makeStyles, Theme } from '@material-ui/core'
import { ChatSharp } from '@material-ui/icons'
import clsx from 'clsx'
import React from 'react'
import { useDocumentPermissions, useLocalization, useViewerState } from '../../hooks'
import { ToggleBase } from './toggle-base'

const useStyles = makeStyles((theme: Theme) => {
  return createStyles({
    iconButton: {},
    icon: {},
    iconActive: {
      fill: theme.palette.primary.main,
    },
  })
})

type AddAnnnotationClassKey = Partial<ReturnType<typeof useStyles>>

/**
 * Document widget component to draw Annotation
 */
export const AddAnnotationWidget: React.FC<{ classes?: AddAnnnotationClassKey }> = (props) => {
  const classes = useStyles(props)
  const localization = useLocalization()
  const viewerState = useViewerState()
  const permissions = useDocumentPermissions()

  const pageRotation =
    viewerState.rotation?.find((rotation) => rotation.pageNum === viewerState.activePage)?.degree || 0

  return (
    <ToggleBase
      disabled={!permissions.canEdit || pageRotation !== 0}
      classes={classes}
      active={viewerState.activeShapePlacing === 'annotation'}
      title={localization.addAnnotation}
      setValue={(value) => viewerState.updateState({ activeShapePlacing: value ? 'annotation' : 'none' })}>
      <ChatSharp
        className={clsx(classes.icon, { [classes.iconActive]: viewerState.activeShapePlacing === 'annotation' })}
      />
    </ToggleBase>
  )
}
