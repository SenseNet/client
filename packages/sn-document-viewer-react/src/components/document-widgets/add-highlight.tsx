import { createStyles, makeStyles, Theme } from '@material-ui/core'
import ChatBubbleOutlineSharp from '@material-ui/icons/ChatBubbleOutlineSharp'
import clsx from 'clsx'
import React from 'react'
import { useDocumentPermissions, useLocalization, useViewerState } from '../../hooks'
import { ToggleBase } from './toggle-base'

const useStyles = makeStyles((theme: Theme) => {
  return createStyles({
    iconButton: {},
    icon: {},
    iconActive: {
      fill: theme.palette.primary.main,
    },
  })
})

type AddRHightlightClassKey = Partial<ReturnType<typeof useStyles>>

/**
 * Document widget component to draw Hihgligh
 */
export const AddHighlightWidget: React.FC<{ classes?: AddRHightlightClassKey }> = (props) => {
  const classes = useStyles(props)
  const localization = useLocalization()
  const viewerState = useViewerState()
  const permissions = useDocumentPermissions()

  return (
    <ToggleBase
      disabled={!permissions.canEdit}
      classes={classes}
      isVisible={viewerState.isPlacingHighlight}
      title={localization.addHighlight}
      setValue={(value) => viewerState.updateState({ isPlacingHighlight: value })}>
      <ChatBubbleOutlineSharp
        className={clsx(classes.icon, { [classes.iconActive]: viewerState.isPlacingHighlight })}
      />
    </ToggleBase>
  )
}
