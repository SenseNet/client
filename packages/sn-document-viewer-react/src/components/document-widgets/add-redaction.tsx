import { createStyles, makeStyles, Theme } from '@material-ui/core'
import ChatBubbleSharp from '@material-ui/icons/ChatBubbleSharp'
import clsx from 'clsx'
import React from 'react'
import { useDocumentPermissions, useLocalization, useViewerState } from '../../hooks'
import { ToggleBase } from './toggle-base'

const useStyles = makeStyles((theme: Theme) => {
  return createStyles({
    iconButton: {},
    icon: {},
    iconActive: {
      fill: theme.palette.primary.main,
    },
  })
})

type AddRedactionClassKey = Partial<ReturnType<typeof useStyles>>

/**
 * Document widget component to toggleing redaction
 */
export const AddRedactionWidget: React.FC<{ classes?: AddRedactionClassKey }> = (props) => {
  const classes = useStyles(props)
  const localization = useLocalization()
  const viewerState = useViewerState()
  const permissions = useDocumentPermissions()

  return (
    <ToggleBase
      disabled={!permissions.canEdit}
      classes={classes}
      isVisible={viewerState.isPlacingRedaction}
      title={localization.addRedaction}
      setValue={(v) => viewerState.updateState({ isPlacingRedaction: v })}>
      <ChatBubbleSharp className={clsx(classes.icon, { [classes.iconActive]: viewerState.isPlacingRedaction })} />
    </ToggleBase>
  )
}
