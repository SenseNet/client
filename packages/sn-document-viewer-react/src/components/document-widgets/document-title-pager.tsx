import ClickAwayListener from '@material-ui/core/ClickAwayListener'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import React, { useState } from 'react'
import { useDocumentData, useLocalization, useViewerState } from '../../hooks'

/**
 * Document widget component for paging
 */
export const DocumentTitlePager: React.FC = () => {
  const { documentData } = useDocumentData()
  const viewerState = useViewerState()
  const localization = useLocalization()
  const [isFocused, setIsFocused] = useState(false)
  const [currentValue, setCurrentValue] = useState<number | string>(viewerState.activePage)

  const gotoPage = (page: string | number) => {
    let pageInt = typeof page === 'string' ? parseInt(page, 10) : page
    if (!isNaN(pageInt)) {
      pageInt = Math.max(pageInt, 1)
      pageInt = Math.min(pageInt, documentData.pageCount)
      viewerState.pageToGo.setValue({ page: pageInt })
    }
  }

  return (
    <ClickAwayListener onClickAway={() => setIsFocused(false)}>
      <Typography
        onClick={() => setIsFocused(true)}
        variant="h6"
        color="inherit"
        style={{ cursor: 'pointer', display: 'flex', alignItems: 'center', overflow: 'hidden', margin: '0 2.5em' }}
        title={documentData.documentName}>
        <div style={{ whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis' }}>
          {documentData.documentName}&nbsp;
        </div>
        {isFocused ? (
          <form
            onSubmit={(ev) => {
              gotoPage(currentValue)
              ev.preventDefault()
            }}>
            <TextField
              style={{ flexShrink: 0 }}
              title={localization.gotoPage}
              onChange={(ev) => setCurrentValue(ev.currentTarget.value)}
              onBlur={() => gotoPage(currentValue)}
              defaultValue={viewerState.activePage}
              type="number"
              required={true}
              InputLabelProps={{
                shrink: true,
              }}
              inputProps={{
                min: 1,
                max: documentData.pageCount,
                style: { textAlign: 'center' },
              }}
              margin="dense"
            />
          </form>
        ) : (
          <div style={{ flexShrink: 0 }}>
            {viewerState.activePage} / {documentData.pageCount}
          </div>
        )}
      </Typography>
    </ClickAwayListener>
  )
}
