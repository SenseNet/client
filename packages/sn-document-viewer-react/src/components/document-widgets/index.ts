export * from './add-annotation'
export * from './add-highlight'
export * from './add-redaction'
export * from './document-title-pager'
export * from './download-widget'
export * from './print-widget'
export * from './rotate-active-pages'
export * from './rotate-document'
export * from './rotate-widget'
export * from './save-widget'
export * from './share-widget'
export * from './toggle-base'
export * from './toggle-comments'
export * from './toggle-redaction'
export * from './toggle-shapes'
export * from './toggle-thumbnails'
export * from './toggle-watermark'
export * from './zoom-in-out'
