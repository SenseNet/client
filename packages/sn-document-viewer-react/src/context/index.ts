export * from './api-settings'
export * from './comment-states'
export * from './comments'
export * from './document-data'
export * from './document-permissions'
export * from './localization-context'
export * from './preview-image-data'
export * from './viewer-settings'
export * from './viewer-state'
