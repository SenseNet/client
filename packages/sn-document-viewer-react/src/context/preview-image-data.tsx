import React, { useCallback, useEffect, useState } from 'react'
import { PreviewImageData } from '@sensenet/client-core'
import {
  useDocumentData,
  useDocumentPermissions,
  useDocumentViewerApi,
  useViewerSettings,
  useViewerState,
} from '../hooks'
import { applyShapeRotations, ImageUtil } from '../services'

export const PreviewImageDataContext = React.createContext<{
  imageData: PreviewImageData[]
  rotateImages: (indexes: number[], amount: number) => void
}>({ imageData: [], rotateImages: () => undefined })

export const PreviewImageDataContextProvider: React.FC = (props) => {
  const viewerSettings = useViewerSettings()
  const api = useDocumentViewerApi()
  const { documentData, updateDocumentData } = useDocumentData()
  const viewerState = useViewerState()
  const permissions = useDocumentPermissions()
  const [previewImages, setPreviewImages] = useState<PreviewImageData[]>([])

  useEffect(() => {
    const abortController = new AbortController()
    ;(async () => {
      try {
        setPreviewImages([])
        const images = await api.getExistingPreviewImages({
          document: documentData,
          version: viewerSettings.version || '',
          showWatermark: viewerState.showWatermark,
          abortController,
        })
        setPreviewImages(images)
      } catch (error) {
        if (!abortController.signal.aborted) {
          throw error
        }
      }
    })()

    return () => abortController.abort()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    api,
    documentData.idOrPath,
    documentData.hostName,
    viewerSettings.documentIdOrPath,
    viewerSettings.version,
    viewerState.showWatermark,
  ])

  const rotateImages = useCallback(
    (imageIndexes: number[], amount: number) => {
      if (!permissions.canEdit) {
        console.warn(`No permission to edit!`)
        return
      }
      const newImages = previewImages.map((img) => {
        const newImg = { ...img }
        if (imageIndexes.indexOf(newImg.Index) >= 0) {
          const newAngle =
            ImageUtil.normalizeDegrees(((newImg.Attributes && newImg.Attributes.degree) || 0) + (amount % 360)) % 360
          newImg.Attributes = {
            ...newImg.Attributes,
            degree: newAngle,
          }
        }

        updateDocumentData({
          shapes: {
            annotations: applyShapeRotations(documentData.shapes.annotations, amount, img),
            highlights: applyShapeRotations(documentData.shapes.highlights, amount, img),
            redactions: applyShapeRotations(documentData.shapes.redactions, amount, img),
          },
        })

        return newImg
      })
      setPreviewImages(newImages)
      viewerState.updateState({ hasChanges: true })
    },
    [documentData, updateDocumentData, permissions.canEdit, previewImages, viewerState],
  )

  return (
    <PreviewImageDataContext.Provider
      value={{
        imageData: previewImages,
        rotateImages,
      }}>
      {props.children}
    </PreviewImageDataContext.Provider>
  )
}
