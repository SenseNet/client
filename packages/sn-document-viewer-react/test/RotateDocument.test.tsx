import { mount, shallow } from 'enzyme'
import React from 'react'
import { RotateDocumentWidget } from '../src/components/document-widgets/RotateDocument'
import { PreviewImageDataContext } from '../src/context/preview-image-data'
import { examplePreviewImageData } from './__Mocks__/viewercontext'

describe('RotateDocument component', () => {
  it('Should render without crashing', () => {
    const rotateImages = jest.fn()
    const wrapper = shallow(
      <PreviewImageDataContext.Provider value={{ imageData: [examplePreviewImageData], rotateImages }}>
        <RotateDocumentWidget />
      </PreviewImageDataContext.Provider>,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('RotateLeft should trigger a rotate to left', () => {
    const rotateImages = jest.fn()
    const wrapper = mount(
      <PreviewImageDataContext.Provider value={{ imageData: [examplePreviewImageData], rotateImages }}>
        <RotateDocumentWidget />
      </PreviewImageDataContext.Provider>,
    )
    wrapper
      .find('#RotateLeft')
      .first()
      .simulate('click')
    expect(rotateImages).toBeCalledWith([1], -90)
  })

  it('RotateRight should trigger a rotate to right', () => {
    const rotateImages = jest.fn()
    const wrapper = mount(
      <PreviewImageDataContext.Provider value={{ imageData: [examplePreviewImageData], rotateImages }}>
        <RotateDocumentWidget />
      </PreviewImageDataContext.Provider>,
    )
    wrapper
      .find('#RotateRight')
      .first()
      .simulate('click')
    expect(rotateImages).toBeCalledWith([1], 90)
  })
})
