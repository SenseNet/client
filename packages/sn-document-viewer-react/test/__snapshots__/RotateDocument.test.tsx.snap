// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RotateDocumentWidget component Should render without crashing 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
    }
  }
>
  <WithStyles(IconButton)
    title="rotateDocumentLeft"
  >
    <pure(RotateLeftIcon)
      onClick={[Function]}
      style={
        Object {
          "border": "2px solid",
          "borderRadius": "5px",
        }
      }
    />
  </WithStyles(IconButton)>
  <WithStyles(IconButton)
    title="rotateDocumentRight"
  >
    <pure(RotateRightIcon)
      onClick={[Function]}
      style={
        Object {
          "border": "2px solid",
          "borderRadius": "5px",
        }
      }
    />
  </WithStyles(IconButton)>
</div>
`;
